File: MissingInteger.cpp
// you can use includes, for example:
// #include <algorithm>
#include <set>
// you can write to stdout for debugging purposes, e.g.
// cout << "this is a debug message" << endl;

int solution(vector<int> &A) {
    // write your code in C++11 (g++ 4.8.2)
    if (A.size() == 0)
        return 1;
    
    int min_a = A[0];
    int max_a = A[0];
    set<int> B = {};
    
    for (auto i: A)
    {
        min_a = min(i, min_a);
        max_a = max(i, max_a);
        if (i>0)
            B.insert(i);
    }
    int start = 1;
    int end = B.size()+2;
    while (start < end)
    {
        if (B.find(start) == B.end())
            break;
        start++;
    }
    return start;
}

File: Counting Elements/MissingInteger.py
def solution(A):
    # write your code in Python 2.7
    import sys
    min_a = sys.maxint
    max_a = -min_a - 1
    
    from sets import Set
    B=Set()
    
    if len(A)==0:
        return 1
    
    for i in A:
        if i>0:
            min_a = min(min_a,i)
            max_a = max(max_a,i)
            B.add(i)
    
    if max_a < min_a or min_a<0 or max_a < 0:
        return 1
    
    ret = 1
    start =1
    while start < min(max_a+2,sys.maxint):
        if start not in B:
            ret = start
            break
        start = start + 1
    
    return ret